### ISSUE TITLE
Twitch showing ads/commercials again

### ISSUE THREAD

**Issue by Abalieno** [streamlink/streamlink#3120]
## Bug Report

- [X] This is a bug report and I have read the contribution guidelines.

### Description

Using: --player-passthrough hls --twitch-disable-ads
Since early today I noticed that the player started again showing ads interruptions. Not the ads themselves but a screen replacing them.

By the way, I noticed someone else opened a similar issue. This is new. I was using the command line the last few months. Only today I started seeing this screen. I've also updated Streamlink to the last version and the problem persists.

**Comment by bastimeyer** [comment]
You can't use `--player-passthrough hls`, as it will completely circumvent Streamlink's own HLS client, hence the name "passthrough". All it does is launch your player with the resolved HLS stream URL from Twitch as the first parameter, which means your player will read the HLS stream on its own, and not Streamlink's stream output like usual. `--twitch-disable-ads` is an extension to Streamlink's HLS client, so when using it, the output of Streamlink will strip off ad segments.

Also, this thread is a duplicate of #3119 from just a few minutes ago. Instead of closing it, I will keep it open for now and lock it, so that people who don't look for already existing "issues" don't create any more duplicates, because I feel like that's going to happen.

**Comment by Abalieno** [comment]
I've just tested it with only --twitch-disable-ads, it still shows the ads screen.

I'm also pretty sure I had to add the passthrough command last time to make it work. In any case, removing the passthrough option and keeping --twitch-disable-ads doesn't remove the ads screen.

**Comment by bastimeyer** [comment]
> I'm also pretty sure I had to add the passthrough command last time to make it work.

No, wrong. As I've already said,
```
streamlink --player "$player" --player-passthrough=hls "$input" "$quality"
```
translates to
```
"$player" "$(streamlink --stream-url "$input" "$quality")"
```
And unless your player can filter out ad segments from an HLS Twitch stream on its own, it won't work.

> In any case, removing the passthrough option and keeping --twitch-disable-ads doesn't remove the ads screen.

Post a debug log, as mentioned in the other thread.
https://github.com/streamlink/streamlink/issues/3119#issuecomment-676812980

----

> The current behavior is that it waits the ads to end before starting the stream. I'll update in case I see ads interruptions again during the stream

And there's no way around that, since the ads are embedded.
You won't have to wait, because it supports filtering out ads at the beginning (pre-roll) and regualr mid-stream ads. When they get filtered out, the stream will have a gap, so it's possible that you'll need to set the `--player-no-close` parameter as well, in case your player decides to terminate its process when there's no input to read from for a short while.

**Comment by Abalieno** [comment]
What I meant is that since I've added both passthrough and no-ads it has been working correctly and also filtering the embedded screens. Something changed today.

It does seem to wait the same 15 seconds of the initial ads screen before starting. I'll see if I spot interruptions mid-stream.

**Comment by kuon** [comment]
I also noticed a change today, now streamlink waits like 20 seconds before launching the stream, with this message:

`[plugin.twitch][info] Waiting for pre-roll ads to finish, be patient`

I never had that before, that's why I think something might have changed.

**Comment by gravyboat** [comment]
@bastimeyer We're probably going to see an influx of these issues because I also got the pre-roll ad notification on Monday, but not yesterday, and then I got them it again today. I don't have `twitch-disable-ads` set on my machine and wasn't getting pre-roll notifications since Twitch last messed around with them so they're probably just re-enabling. I think leaving this issue open is good, but let's clarify the following for anyone else who:

**There is nothing Streamlink or the Streamlink dev team can do about pre-roll ad notification embedded in the stream itself. This is not something new Twitch is doing, they enable and disable the embedded notification from time to time, we have no control or impact on this. It is completely on Twitch. You can use `--twitch-disable-ads` if you wish and Streamlink will not launch your player until the embed finishes, or you can just wait, these are the only options. Adding additional comments to this thread is pointless and unnecessary.**

**Comment by jimbo1111** [comment]
I am having this issue, and I recognize the last comment saying it's useless to add stuff to this thread, but I thought I would add that this is affecting all streams, including ones that I have a subscription to (and thus would not normally show ads).  When I try watching the ads in any browser, the pre-roll ads don't show up (though it does show the browser connecting to Twitch's ad system, and then not running a pre-roll).  If disable ads is on, it will tell me to wait.  If disable ads is off, it will load the stream and show a 15 second countdown "waiting for your ad to finish" or something like that.  Not that I know anything about programming, but it does feel like perhaps it would be possible to bypass that 15 second wait, somehow...

**Comment by gravyboat** [comment]
@jimbo1111 That's because we dropped support for being able to log in to Twitch via Streamlink so your subscription status isn't recognized (we had to do this for some breaking changes a while back). In terms of being able to bypass that 15 second wait it's impossible because it's actually inserted into the stream itself, there's no way for us to bypass it because it's showing that to new connections.

**Comment by bastimeyer** [comment]
Since more comments will further bury this, like it always does when Twitch re-enables embedded ads or makes other changes, I'm going to lock the thread now.

`--twitch-disable-ads` is working as intended and it will filter out the ads correctly. Pre-roll ads will delay opening the player, as no stream data is available until the ads have finished. To address the confusion of some users, you **cannot** use `--player-passthrough=hls` or similar options with that, as explained above. Won't repeat this again...

And NO, you cannot circumvent **embedded ads**, as they replace the regular content, similar to ads while watching a live event on TV. When Twitch decides to show embedded ads to everyone on any stream, then you will have to accept that unless there is a *magic* request parameter to tell them to not do that (which there isn't). Whatever Twitch is doing on their website with their web-player and external ads-system is irrelevant to the HLS stream Streamlink reads. This has been discussed plenty of time already in various other threads in the past months/years. And subscriptions are irrelevant either, because Twitch made changes at the end of last year which required us and every other 3rd party Twitch API developer to remove all Twitch login/signin functionalities when aquiring the stream access token, needed for watching streams.

----

Also, to quote my comments from earlier:

https://github.com/streamlink/streamlink/issues/3119#issuecomment-676811866
https://github.com/streamlink/streamlink/issues/3119#issuecomment-676812980

> This is nothing new. Twitch has implemented and run embedded ads several times now, once even breaking our old implementation for filtering out ads which had to be rewritten. **They don't always have embedded ads active, which is the reason why you've not been seeing this recently.**

> [plugin.twitch][info] Will skip ad segments
> [plugin.twitch][info] Waiting for pre-roll ads to finish, be patient
> [cli][info] Starting player: mpv

```
$ streamlink --loglevel debug --twitch-disable-ads twitch.tv/gamesdonequick best
[cli][debug] OS:         Linux-5.8.0-1-git-x86_64-with-glibc2.2.5
[cli][debug] Python:     3.8.5
[cli][debug] Streamlink: 1.5.0+10.g00fd4d2
[cli][debug] Requests(2.24.0), Socks(1.7.1), Websocket(0.56.0)
[cli][info] Found matching plugin twitch for URL twitch.tv/gamesdonequick
[cli][debug] Plugin specific arguments:
[cli][debug]  --twitch-disable-ads=True (disable_ads)
[plugin.twitch][debug] Getting live HLS streams for gamesdonequick
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: audio_only, 160p (worst), 360p, 480p, 720p, 720p60, 1080p60 (best)
[cli][info] Opening stream: 1080p60 (hls)
[plugin.twitch][info] Will skip ad segments
[stream.hls][debug] Reloading playlist
[plugin.twitch][info] Waiting for pre-roll ads to finish, be patient
[stream.hls][debug] First Sequence: 0; Last Sequence: 2
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 0; End Sequence: None
[stream.hls][debug] Adding segment 0 to queue
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Adding segment 1 to queue
[stream.hls][debug] Adding segment 2 to queue
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 3 to queue
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 4 to queue
[stream.hls][debug] Adding segment 5 to queue
[stream.hls][debug] Adding segment 6 to queue
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 7 to queue
[stream.hls][debug] Adding segment 8 to queue
[stream.hls][debug] Download of segment 8 complete
[cli][info] Starting player: mpv
[cli.output][debug] Opening subprocess: mpv --title=twitch.tv/gamesdonequick --cache=no -
[cli][debug] Writing stream to output
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Adding segment 9 to queue
[stream.hls][debug] Adding segment 10 to queue
[stream.hls][debug] Adding segment 11 to queue
[stream.hls][debug] Download of segment 9 complete
[cli][info] Player closed
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

**Comment by bastimeyer** [comment]
I've reopened the thread and pinned it again for visibility purposes, because people are too lazy to search for already existing threads and thus spam the issue tracker. I hope this will reduce the spamming.

Also, if you're using the `--twitch-disable-ads` parameter in additon to `--twitch-low-latency`, it will incorrectly print a "This is not a low latency stream" info message, because it can't find any prefetch segments in the pre-roll ads playlist. This message needs to be fixed/improved, but it doesn't affect the actual low latency streaming logic.

**Comment by bastimeyer** [comment]
A quick update regarding the embedded ads on Twitch.

The "magic" request parameter which I was talking about earlier that "doesn't exist" seems to be working again. I say "again", because when Twitch first implemented the embedded ads, adding the `platform=_` GET parameter when aquiring the access token made it possible to prevent ads from being included and embedded into the HLS stream. This request parameter however only kept working for a few days and Twitch made further changes, which eventually lead to the implementation of the `--twitch-disable-ads` parameter in Streamlink. After that, the `platform=_` request parameter got removed in another change, as it seemed pointless to keep. As it turns out now, it is still working again and we've added it back (see #3173). It is unknown though how long this will keep working this time, but at least it's working for now.

If you don't want to wait until the next Streamlink release (date unknown), then you can
- build/install Streamlink from the master branch, preferably via python pip
  https://streamlink.github.io/install.html#pypi-package-and-source-code
- or, if you're on Windows, wait for the next development build / nightly installer for Windows to be built and download that (2020-09-15)
  https://streamlink.github.io/install.html#windows-binaries

The incorrect info log message regarding low latency streams will be fixed once #3169 got merged, but with the ads workaround, this currently doesn't make much of a difference.

**Comment by bastimeyer** [comment]
**To prevent ads from being embedded into Twitch streams, install the latest release (1.6.0)**
- https://streamlink.github.io/install.html
- https://github.com/streamlink/streamlink/releases/latest
- https://github.com/streamlink/streamlink/blob/master/CHANGELOG.md#streamlink-160-2020-09-22

Let's keep this thread open and pinned for a couple more days.

**Comment by bastimeyer** [comment]
Forgot to close this, as we have a new sticky thread since a couple of days.
